### ISSUE TITLE
NTv2 grid files with geoid

### ISSUE THREAD

**Issue by dfileccia** [OSGeo/PROJ#1315]
I am using PROJ.4 6.0.0 and I am surprised full support for NTv2 files has not been implemented at this time. I am not a surveyor but I need coordinate transformations and it would be nice to not distribute several file formats for the areas of the world that do not support standard files.

The UK Ordnance Survey OSTN02 and OSTN15 files contain both the horizontal and vertical grid shifts. These are GSB files and there are no GTX file for geoids based on this data. It would be nice if PROJ.4 would use GSB data if it is present.

An example could be:
+proj=tmerc ... +grids=@OSTN15_NTv2_OSGBtoETRS.gsb

Notice it is not using geoidgrids or nadgrids. By doing so the file could be used for both.

**Comment by kbevers** [comment]
> I am using PROJ.4 6.0.0 and I am surprised full support for NTv2 files has not been implemented at this time.

It is. The grid you are using is even included in [`proj-datumgrid-europe`](https://github.com/OSGeo/proj-datumgrid/blob/eca9aa839c9ca1b812e759ddac42044f6ba983ef/europe/README.EUROPE#L87-L101).

You *can* use the deprecated `+nadgrids` flags with other grids than NAD27/83 datumshift grids. The modern way to it is better:

```
+proj=pipeline +step +proj=hgridshift +grids=OSTN15_NTv2_OSGBtoETRS.gsb +step +proj=tmerc ...
```

Have a look at the documentation for more info: https://proj4.org/operations/transformations/hgridshift.html.

**Comment by dfileccia** [comment]
Hi Kristian,

I am fully aware  of the horizontal shift aspects. I used it in 5.0 as well (with this file). Unfortunately, the NTv2 also contains the vertical shifts too. PROJ.4 "ONLY" supports the GTX format for vertical datums. The OSTN15 files are GSB format which is supported by the horizontal shifts.

I am asking that PROJ.4 handle BOTH the horizontal and vertical shifts from the GSB file. This seems to be a lack in feature. The UK Ordnance Survey publishes a document on how to do just that.

So, any solution? Not according to the source code at this point.

**Comment by kbevers** [comment]
> Unfortunately, the NTv2 also contains the vertical shifts too. 

Does it really? The [Ordnance Survey has this to say about the matter](https://www.ordnancesurvey.co.uk/blog/2016/09/ostn15-new-geoid-britain/):

> This is the case in Britain and we have a transformation model between ETRS89 and OSGB36 called OSTN15 and a separate model for ETRS89 height to ODN called OSGM15. When our surveyors are out using OS Net to update our maps their GPS coordinates are transformed to OSGB36 and ODN through OSTN15/OSGM15.

To me it sounds like you need a model for the horizontal part and one for the vertical part. Which aligns quite well with how this is done in other places in the world.

> The UK Ordnance Survey publishes a document on how to do just that.

Can you supply a link to that document? 

It is news to me that NTv2 files can also contain the vertical component. As far as I can tell the [original specification](https://web.archive.org/web/20091227232322/http://www.mgs.gov.on.ca/stdprodconsume/groups/content/@mgs/@iandit/documents/resourcelist/stel02_047447.pdf) of the format doesn't mention a vertical component. GDAL doesn't know about that dimension either:

```
$ gdalinfo OSTN15_NTv2_OSGBtoETRS.gsb
Driver: NTv2/NTv2 Datum Grid Shift
Files: OSTN15_NTv2_OSGBtoETRS.gsb
Size is 661, 1441
Coordinate System is:
GEOGCS["WGS 84",
    DATUM["WGS_1984",
        SPHEROID["WGS 84",6378137,298.257223563,
            AUTHORITY["EPSG","7030"]],
        AUTHORITY["EPSG","6326"]],
    PRIMEM["Greenwich",0,
        AUTHORITY["EPSG","8901"]],
    UNIT["degree",0.0174532925199433,
        AUTHORITY["EPSG","9122"]],
    AUTHORITY["EPSG","4326"]]
Origin = (-9.008333333333333,61.004166666666670)
Pixel Size = (0.016666666666667,-0.008333333333333)
Metadata:
  CREATED=20161003
  GS_TYPE=SECONDS
  MAJOR_F=6377563.396
  MAJOR_T=6378137
  MINOR_F=6356256.909
  MINOR_T=6356752.314
  PARENT=NONE
  SUB_NAME=GB+IOM
  SYSTEM_F=AIRY1830
  SYSTEM_T=GRS80

  UPDATED=20161003
  VERSION=OSTN15
Subdatasets:
  SUBDATASET_0_NAME=NTv2:0:OSTN15_NTv2_OSGBtoETRS.gsb
  SUBDATASET_0_DESC=GB+IOM
Corner Coordinates:
Upper Left  (  -9.0083333,  61.0041667) (  9d 0'30.00"W, 61d 0'15.00"N)
Lower Left  (  -9.0083333,  48.9958333) (  9d 0'30.00"W, 48d59'45.00"N)
Upper Right (   2.0083333,  61.0041667) (  2d 0'30.00"E, 61d 0'15.00"N)
Lower Right (   2.0083333,  48.9958333) (  2d 0'30.00"E, 48d59'45.00"N)
Center      (  -3.5000000,  55.0000000) (  3d30' 0.00"W, 55d 0' 0.00"N)
Band 1 Block=661x1 Type=Float32, ColorInterp=Undefined
  Description = Latitude Offset (arc seconds)
Band 2 Block=661x1 Type=Float32, ColorInterp=Undefined
  Description = Longitude Offset (arc seconds)
Band 3 Block=661x1 Type=Float32, ColorInterp=Undefined
  Description = Latitude Error
Band 4 Block=661x1 Type=Float32, ColorInterp=Undefined
  Description = Longitude Error
```

> So, any solution? Not according to the source code at this point.

At this point I am not sure there even is a problem, but if what you are saying is correct then there isn't an immediate solution, except from splitting into two grids.

**Comment by dfileccia** [comment]
Thanks Kristian,

I have done the above too. I cannot seem to find an OSGM15 files for the geoid.

Here is the link I have been using:
https://www.ordnancesurvey.co.uk/business-and-government/help-and-support/navigation-technology/os-net/formats-for-developers.html

"Developer's Pack"

And, I did misread their PDF files. I assumed their correction explanation, since it used shifts to calculate the geoid area, was all in the same file. This was from the OS MasterMap Site Layer User Guide (page 22). As I said, I am not a surveyor (just an electrical engineer) and most of this stuff is greek to me. I do understand the need for spatial referencing though.

If you can point me to OSGM15 data files it would be nice. I already use Belfast and Malin but I have none for GB.

**Comment by kbevers** [comment]
>  I cannot seem to find an OSGM15 files for the geoid.

I am in the same boat. I assume you have to use the software provided by the Ordnance Survey to get access to the geoid somehow. It may be that you can get a better answer if you ask about this topic on the PROJ mailing list. Or if you contact the Ordnance Survey and have them explain it better (if you do, please share your results with us).

**Comment by kbevers** [comment]
I did some more research. The geoid grid is contained in "developers pack" in the `OSTN15_OSGM15_DataFile.txt` file. It's a text file that needs to be converted to the GTX format before it can be used in PROJ. The `Transformations_and_OSGM15_User_Guide.pdf` file explains how to read the grid. It should be fairly easy to convert to a different format. Should you choose to do so we are happy to accept it as part of https://github.com/OSGeo/proj-datumgrid.

I am closing this issue for now, as I think we have covered the topic sufficiently at this point.
